src/transaction.rs

```diff
@@ -303,12 +304,84 @@ impl Transaction {
 
     pub fn set_receipts_root(&mut self, root: Bytes32) -> Option<Bytes32> {
         match self {
-            Self::Script { receipts_root, .. } => Some(std::mem::replace(receipts_root, root)),
+            Self::Script { receipts_root, .. } => Some(mem::replace(receipts_root, root)),
 
             _ => None,
         }
     }
 
+    /// Append a new unsigned input to the transaction.
+    ///
+    /// When the transaction is constructed, [`Transaction::sign_inputs`] should
+    /// be called for every secret key used with this method.
+    ///
+    /// The production of the signatures can be done only after the full
+    /// transaction skeleton is built because the input of the hash message
+    /// is the ID of the final transaction.
+    #[cfg(feature = "std")]
+    pub fn add_unsigned_coin_input(
+        &mut self,
+        utxo_id: UtxoId,
+        owner: &PublicKey,
+        amount: Word,
+        color: Color,
+        maturity: Word,
+        predicate: Vec<u8>,
+        predicate_data: Vec<u8>,
+    ) {
+        let owner = Input::coin_owner(owner);
+
+        let witness_index = self.witnesses().len() as u8;
+        let input = Input::coin(
+            utxo_id,
+            owner,
+            amount,
+            color,
+            witness_index,
+            maturity,
+            predicate,
+            predicate_data,
+        );
+
+        self._add_witness(Witness::default());
```

Comment:
This seems inefficient because all of these witnesses would be duplicates if all coins have the same owner field. One major advantage of the witness index is that it allows inputs to share common witnesses, instead of requiring a witness on the coin input type.

Reply:
That is true. This will involve some state tracking in the transaction builder to construct the witnesses in an optimal way.

The current implementation won't affect the correctness of the protocol. We can create this improvement as a separate PR/issue: https://github.com/FuelLabs/fuel-tx/issues/72

Reply:
Also I think the sdk won't be using this, so it's not urgent.